torch/onnx/_internal/exporter/_core.py

```diff
@@ -794,13 +957,60 @@ def _exported_program_to_onnx_program(
         producer_version=torch.__version__,
     )
 
-    if lower == "none":
-        # Add the opset import for the torch ops
-        model.opset_imports["pkg.torch.ops"] = _torch_version_integer()
-    # NOTE: Function domains are added when translating nodes when lower="at_conversion"
+    # Opset imports are added to the model in the final add_opset_imports pass
 
-    # 1. Add all nodes to the graph and create a dictionary of values
-    values = _add_nodes(exported_program, model, lower=lower, registry=registry)
+    # A dictionary storing the translated subgraphs as ONNX functions made available to outer graphs
+    # {<subgraph_scope>: {<subgraph_name>: <IR function>}}
+    scoped_subgraphs: dict[str, dict[str, ir.Function]] = {}
+    values = None
+
+    # 1. Translate all nodes in all subgraphs and the main graph
+    # Create a dictionary of values for the main graph for step 2-3 to add inputs and outputs
+    module: torch.fx.GraphModule
+    # Reverse the order of the modules so that the innermost module is processed first
+    # and made available to the outer module
+    for name, module in reversed(
+        tuple(exported_program.graph_module.named_modules(remove_duplicate=False))
+    ):
+        # Obtain the graphs (previously built) owned by the current module
+        owned_graphs = scoped_subgraphs.setdefault(name, {})
+        fx_graph = module.graph
+
+        graph_like: ir.Graph | ir.Function
+        if name == "":
+            # Root graph
+            graph_like = model.graph
+        else:
+            function_name = name.replace(".", "__")
+            # Inputs and outputs will be created within _translate_fx_graph
+            func = ir.Function(
+                domain=_LOCAL_FUNCTION_DOMAIN,
+                name=function_name,
+                graph=ir.Graph((), (), nodes=()),
+                attributes=(),
+            )
+            # Make this function available to the outer graph
+            scope, subgraph_name = _get_scope_name(name)
+            scoped_subgraphs.setdefault(scope, {})[subgraph_name] = func
+            model.functions[func.identifier()] = func
```

Comment:
Should we have a method for that? That way, we can check it is not overwriting an existing one.

Reply:
Do you think an assert will do? Or do you think we should deduplicate?

Reply:
The function names should be unique: given that the names are scoped and unique (e.g. `scope1.scope2.module_name` when calling get_modules(), the replacement we do here is just `"."->`"__"` to make it a valid ONNX name.